### PR TITLE
deprovision and delete agent code stops and delete microservices

### PR DESCRIPTION
**Pull request by NehaNaithani**

### CODE REVIEW

**Review by baghbidi — COMMENTED** [review]
Looks good to me. However I doubt all the microservices get stopped after running `iofogctl delete agent ...`. because that command stops the daemon right after deprovisioning the agent and agent never gets a chance to process STOP tasks that are on the queue.

File: iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/ProcessManager.java [hunk 1]
```diff
@@ -239,16 +240,80 @@ public void deleteRemainingMicroservices() {
...
 
-	private void deleteOldAgentContainers(Set<String> oldAgentContainerNames) {
-		logInfo("Delete old agent containers");
-		oldAgentContainerNames.forEach(name -> {
+	/**
+	 * Stop running microservices when agent deprovision.
+	 * Stop and delete microservices when agents stops
+	 * @param withCleanUp
+	 */
+	public void stopRunningMicroservices(boolean withCleanUp) {
```
Comment (baghbidi) [inline iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/ProcessManager.java]: If agent is deployed as a docker container (which we can deploy multiple agents on a machine), this method stops and deletes all microservices, no matter if they belong to the agent that is being stopped.
Check these lines out to see how you can filter those microservices:
https://github.com/eclipse-iofog/Agent/blob/9fc79d1b919397dbef70a4544f8b65f2a78cc491/iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/ProcessManager.java#L207
https://github.com/eclipse-iofog/Agent/blob/9fc79d1b919397dbef70a4544f8b65f2a78cc491/iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/ProcessManager.java#L231

**Review by baghbidi — CHANGES_REQUESTED** [review]

File: iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/ProcessManager.java [hunk 2]
```diff
@@ -239,16 +240,95 @@ public void deleteRemainingMicroservices() {
...
+	 * @param withCleanUp
+	 */
+	public void stopRunningMicroservices(boolean withCleanUp, String iofogUuid) {
+		LoggingService.logInfo(MODULE_NAME ,"Stop delete Remaining Microservices");
+		if (withCleanUp) {
+			List<Container> allContainers;
+			synchronized (deleteLock) {
+				allContainers = docker.getContainers();
+			}
+			Set<String> allMicroserviceUuids = allContainers
```
Comment (baghbidi) [inline iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/ProcessManager.java:255]: Shouldn't we check labels here as well?

Comment (NehaNaithani) [reply iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/ProcessManager.java:255]: In this flow the agent is deprovisioned already and the value of `Configuration.getIofogUuid` is set to blank. Therefore I have skipped the checking of label.

File: iofog-agent-daemon/src/main/java/org/eclipse/iofog/command_line/CommandLineAction.java [hunk 3]
```diff
@@ -53,6 +55,14 @@
 
 		@Override
 		public String perform(String[] args) {
+			try {
```
Comment (baghbidi) [inline iofog-agent-daemon/src/main/java/org/eclipse/iofog/command_line/CommandLineAction.java]: the `try/catch` block should be inside the `if` statement

**Review by baghbidi — APPROVED** [review]
Left a comment, looks good to me

File: iofog-agent-daemon/src/main/java/org/eclipse/iofog/command_line/CommandLineAction.java (hunk 3 above)
Comment (NehaNaithani) [reply iofog-agent-daemon/src/main/java/org/eclipse/iofog/command_line/CommandLineAction.java]: Updating code.

**Pull request merged**